internal/controller/iprangeclaim_controller.go

```diff
@@ -0,0 +1,323 @@
+func (r *IpRangeClaimReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	logger.Info("reconcile loop started")
+
+	o := &netboxv1.IpRangeClaim{}
+	err := r.Client.Get(ctx, req.NamespacedName, o)
+	if err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// if being deleted
+	if !o.ObjectMeta.DeletionTimestamp.IsZero() {
+		// end loop if deletion timestamp is not zero
+		return ctrl.Result{}, nil
+	}
+
+	//  check if matching IpRange object already exists
+	ipRange := &netboxv1.IpRange{}
+	ipRangeName := o.ObjectMeta.Name
+	ipRangeLookupKey := types.NamespacedName{
+		Name:      ipRangeName,
+		Namespace: o.Namespace,
+	}
+
+	err = r.Client.Get(ctx, ipRangeLookupKey, ipRange)
+	if err != nil {
+		// return error if not a notfound error
+		if !apierrors.IsNotFound(err) {
+			return ctrl.Result{}, err
+		}
+
+		logger.V(4).Info("iprange object matching iprange claim was not found, creating new iprange object")
+
+		ipRangeModel, res, err := r.restoreOrAssignIpRangeAndSetCondition(ctx, o)
+		if ipRangeModel == nil {
+			return res, err
+		}
+
+		// create the IpRange CR
+		ipRangeResource := generateIpRangeFromIpRangeClaim(ctx, o, ipRangeModel.StartAddress, ipRangeModel.EndAddress)
+		err = controllerutil.SetControllerReference(o, ipRangeResource, r.Scheme)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		err = r.Client.Create(ctx, ipRangeResource)
+		if err != nil {
+			errSetCondition := r.logErrorSetConditionAndCreateEvent(ctx, o, netboxv1.ConditionIpRangeAssignedFalse, corev1.EventTypeWarning, "", err)
+			if errSetCondition != nil {
+				return ctrl.Result{}, errSetCondition
+			}
+			return ctrl.Result{}, err
+		}
+
+		err = r.logErrorSetConditionAndCreateEvent(ctx, o, netboxv1.ConditionIpRangeAssignedTrue, corev1.EventTypeNormal,
+			fmt.Sprintf(" , assigned ip range: %s-%s", ipRangeModel.StartAddress, ipRangeModel.EndAddress), nil)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+	} else {
+		// update fields of IPRange object
+		logger.V(4).Info("update iprange resource")
+		err := r.Client.Get(ctx, ipRangeLookupKey, ipRange)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		updatedIpRangeSpec := generateIpRangeSpec(o, ipRange.Spec.StartAddress, ipRange.Spec.EndAddress, logger)
+		_, err = ctrl.CreateOrUpdate(ctx, r.Client, ipRange, func() error {
+			// only add the mutable fields here
+			ipRange.Spec.CustomFields = updatedIpRangeSpec.CustomFields
+			ipRange.Spec.Comments = updatedIpRangeSpec.Comments
+			ipRange.Spec.Description = updatedIpRangeSpec.Description
+			ipRange.Spec.PreserveInNetbox = updatedIpRangeSpec.PreserveInNetbox
+			return nil
+		})
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	if apismeta.IsStatusConditionTrue(ipRange.Status.Conditions, "Ready") {
+		logger.V(4).Info("iprange status ready true")
+		o.Status, err = r.generateIpRangeClaimStatus(o, ipRange)
```

Comment:
there is no call to update after this to store the new status (unless this is achieved by setting the condition, but then this is a 'happy' side effect